File: 2020/15/sol2.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

int speak(const vector<int> &data, int target) {
  unordered_map<int, pair<int, int>> pos;
  for (int i = 0; i < data.size(); ++i) {
    pos[data[i]] = {i + 1, 0};
  }

  int last = data.back();
  for (int cur = data.size() + 1; cur <= target; ++cur) {
    if (pos.find(last) != pos.end()) {
      auto x = pos[last];
      if (x.second == 0) {
        last = 0;
      } else {
        last = x.first - x.second;
      }
    } else {
      last = 0;
    }
    if (pos.find(last) != pos.end()) {
      pos[last] = {cur, pos[last].first};
    } else {
      pos[last] = {cur, 0};
    }
  }
  return last;
}

int main(int argc, char **argv) {
  int size = 30000000;
  if (argc > 1) {
    size = stoi(argv[1]);
  }
  cout << speak({6, 13, 1, 15, 2, 0}, size) << endl;
  return 0;
}

File: 2020/15/sol2.py
#!/usr/bin/env python3

from tqdm import tqdm


def speak(data, target=30000000):
    pos = {num: (idx, 0) for idx, num in enumerate(data, start=1)}
    cur, last = len(data) + 1, data[-1]

    for cur in tqdm(range(4, target + 1)):
        if last in pos:
            first, second = pos[last]
            last = 0 if second == 0 else first - second
        else:
            last = 0
        pos[last] = (cur, pos[last][0] if last in pos else 0)
    return last


if __name__ == "__main__":
    data = list(map(int, open("input").readline().split(",")))
    print(speak(data))
